### ISSUE TITLE
Cannot run veeamtransport - Warning: Unsupported libc Syscall 0x11C (284)

### ISSUE THREAD

**Issue by kaysond** [ptitSeb/box64#965]
I'm trying to run [veeam's](https://www.veeam.com/) transport service binary using box64, since it only supports x86-64 officially. If I just run it without doing anything special, I get the following output:

```
Dynarec for ARM64, with extension: ASIMD CRC32 PageSize:4096 Running on Cortex-A72 with 4 Cores
Params database has 25 entries
Warning, unrecognized option '--verbose'
Box64 with Dynarec v0.2.5 aa051b66 built on Aug 27 2023 06:31:14
Using default BOX64_LD_LIBRARY_PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
Using default BOX64_PATH: ./:bin/
Counted 25 Env var
Looking for /opt/veeam/transport/veeamtransport
argv[1]="--run-service"
Rename process to "veeamtransport"
Using native(wrapped) librt.so.1
Using native(wrapped) libdl.so.2
Using native(wrapped) libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Using native(wrapped) libpthread.so.0
Using native(wrapped) libutil.so.1
Using emulated libacl.so.1
Using native(wrapped) libm.so.6
Using emulated libattr.so.1
Dynarec for ARM64, with extension: ASIMD CRC32 PageSize:4096 Running on Cortex-A72 with 4 Cores
Params database has 25 entries
Box64 with Dynarec v0.2.5 aa051b66 built on Aug 27 2023 06:31:14
Using default BOX64_LD_LIBRARY_PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
Using default BOX64_PATH: ./:bin/
Counted 25 Env var
Looking for /opt/veeam/transport/veeamtransport
argv[1]="--run-environmentsvc"
argv[2]="8:7"
Rename process to "veeamtransport"
Using native(wrapped) librt.so.1
Using native(wrapped) libdl.so.2
Using native(wrapped) libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Using native(wrapped) libpthread.so.0
Using native(wrapped) libutil.so.1
Using emulated libacl.so.1
Using native(wrapped) libm.so.6
Using emulated libattr.so.1
Dynarec for ARM64, with extension: ASIMD CRC32 PageSize:4096 Running on Cortex-A72 with 4 Cores
Params database has 25 entries
Box64 with Dynarec v0.2.5 aa051b66 built on Aug 27 2023 06:31:14
Using default BOX64_LD_LIBRARY_PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
Using default BOX64_PATH: ./:bin/
Counted 25 Env var
Looking for /opt/veeam/transport/veeamimmureposvc
argv[1]="--subprocess"
argv[2]="--log"
argv[3]="/var/log/VeeamBackup"
argv[4]="--stdio"
argv[5]="10:8"
Rename process to "veeamimmureposvc"
Using native(wrapped) librt.so.1
Using native(wrapped) libdl.so.2
Using native(wrapped) libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Using native(wrapped) libpthread.so.0
Using native(wrapped) libutil.so.1
Using native(wrapped) libm.so.6
Warning: Unsupported libc Syscall 0x11C (284)
Can't log to file
POSIX: Received signal.
Can't log to file
POSIX: Received signal.
Function not implemented
[InotifyMonitor] Failed to create descriptor for unblocking
```

It looks like this syscall is unsupported (eventfd)?

I tried emulating it instead using `$BOX64_EMULATED_LIBS`, but it didn't take effect. I guessed that maybe it's looking for the x86-64 version, so I copied it over from a Debian 10 install, and now it emulates, but I get the following segfault:

```
Dynarec for ARM64, with extension: ASIMD CRC32 PageSize:4096 Running on Cortex-A72 with 4 Cores
Params database has 25 entries
Warning, unrecognized option '--verbose'
Box64 with Dynarec v0.2.5 aa051b66 built on Aug 27 2023 06:31:14
Using default BOX64_LD_LIBRARY_PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
BOX64 will force the used of emulated libs for libc.so.6
Using default BOX64_PATH: ./:bin/
Counted 25 Env var
Looking for /opt/veeam/transport/veeamtransport
argv[1]="--run-service"
Rename process to "veeamtransport"
Using native(wrapped) librt.so.1
Using native(wrapped) libdl.so.2
Using emulated libc.so.6
Redirecting overridden malloc function for libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Error: Global Symbol _dl_argv not found, cannot apply R_X86_64_GLOB_DAT @0x1009d1dd8 ((nil)) in libc.so.6
Using emulated libacl.so.1
Using native(wrapped) libm.so.6
Using native(wrapped) libpthread.so.0
Using emulated libattr.so.1
Warning: Global Symbol _Jv_RegisterClasses not found, cannot apply R_X86_64_GLOB_DAT @0x100e03b60 ((nil)) in libattr.so.1
Warning: Global Symbol _Jv_RegisterClasses not found, cannot apply R_X86_64_GLOB_DAT @0x100bf6c68 ((nil)) in libacl.so.1
Error: Global Symbol strcmp not found, cannot apply R_X86_64_GLOB_DAT @0xd9ddc8 ((nil)) in /opt/veeam/transport/veeamtransport
FillBlock triggered a segfault at 0x27020 from 0x34c09aa4
FillBlock at 0x27020 triggered a segfault, canceling
1178709|SIGSEGV @0x34bd1f94 (???(/opt/veeam/transport/veeamtransport+0x34bd1f94)) (x64pc=0x27020/???:"???", rsp=0x1007ff520, stack=0x100000000:0x100800000 own=(nil) fp=0x1007ff528), for accessing 0x27020 (code=1/prot=87), db=(nil)((nil):(nil)/(nil):(nil)/???:clean, hash:0/0) handler=(nil)
RSP-0x20:0x0000000000000000 RSP-0x18:0x0000000000000000 RSP-0x10:0x0000000000000000 RSP-0x08:0x0000000000000000
RSP+0x00:0x0000000000030080 RSP+0x08:0x0000000000000000 RSP+0x10:0x0000000000000002 RSP+0x18:0x00000001007ff725
RAX:0x00000001008af7c0 RCX:0x0000000000000000 RDX:0x00000001007ff550 RBX:0x0000000000000000
RSP:0x00000001007ff520 RBP:0x00000001007ff528 RSI:0x00000001007ff538 RDI:0x0000000000000002
 R8:0x0000000000000000  R9:0x0000000000000000 R10:0x0000000000000000 R11:0x0000000000000000
R12:0x0000000000000000 R13:0x0000000000000000 R14:0x0000000000000000 R15:0x0000000000000000
ES:0x002b CS:0x0033 SS:0x002b DS:0x002b FS:0x0043 GS:0x0053
```

**Comment by ptitSeb** [comment]
Ok, I have added the missing syscall. You can update box64 and try again.

**Comment by kaysond** [comment]
@ptitSeb its missing 284, not 282. Still getting the same error with the latest commit

**Comment by ptitSeb** [comment]
Ah damn... Ok, I added the correct one now.

**Comment by kaysond** [comment]
> Ah damn... Ok, I added the correct one now.

Awesome! That fixed the issue and now two of the processes start up correctly. There's one more that fails due to a misaligned address error caused by libpthread:

```
sudo -H ~/git/box64/build/box64 veeamtransport --run-service
Dynarec for ARM64, with extension: ASIMD CRC32 PageSize:4096 Running on Cortex-A72 with 4 Cores
Params database has 25 entries
Box64 with Dynarec v0.2.5 cf7ac302 built on Sep  9 2023 13:13:10
Using default BOX64_LD_LIBRARY_PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
Using default BOX64_PATH: ./:bin/
Counted 15 Env var
Looking for veeamtransport
argv[1]="--run-service"
Rename process to "veeamtransport"
Using native(wrapped) librt.so.1
Using native(wrapped) libdl.so.2
Using native(wrapped) libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Using native(wrapped) libpthread.so.0
Using native(wrapped) libutil.so.1
Using emulated libacl.so.1
Using native(wrapped) libm.so.6
Using emulated libattr.so.1
Dynarec for ARM64, with extension: ASIMD CRC32 PageSize:4096 Running on Cortex-A72 with 4 Cores
Params database has 25 entries
Box64 with Dynarec v0.2.5 cf7ac302 built on Sep  9 2023 13:13:10
Using default BOX64_LD_LIBRARY_PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
Using default BOX64_PATH: ./:bin/
Counted 15 Env var
Looking for /opt/veeam/transport/veeamtransport
argv[1]="--run-environmentsvc"
argv[2]="8:7"
Rename process to "veeamtransport"
Using native(wrapped) librt.so.1
Using native(wrapped) libdl.so.2
Using native(wrapped) libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Using native(wrapped) libpthread.so.0
Using native(wrapped) libutil.so.1
Using emulated libacl.so.1
Using native(wrapped) libm.so.6
Using emulated libattr.so.1
Dynarec for ARM64, with extension: ASIMD CRC32 PageSize:4096 Running on Cortex-A72 with 4 Cores
Params database has 25 entries
Box64 with Dynarec v0.2.5 cf7ac302 built on Sep  9 2023 13:13:10
Using default BOX64_LD_LIBRARY_PATH: ./:lib/:lib64/:x86_64/:bin64/:libs64/
Using default BOX64_PATH: ./:bin/
Counted 15 Env var
Looking for /opt/veeam/transport/veeamimmureposvc
argv[1]="--subprocess"
argv[2]="--log"
argv[3]="/var/log/VeeamBackup"
argv[4]="--stdio"
argv[5]="10:8"
Rename process to "veeamimmureposvc"
Using native(wrapped) librt.so.1
Using native(wrapped) libdl.so.2
Using native(wrapped) libc.so.6
Using native(wrapped) ld-linux-x86-64.so.2
Using native(wrapped) libpthread.so.0
Using native(wrapped) libutil.so.1
Using native(wrapped) libm.so.6
126703|SIGBUS @0x7f933e16fc (???(/lib/aarch64-linux-gnu/libpthread.so.0+0x7f933e16fc)) (x64pc=0x800f3/???:"???", rsp=0x1007fe548, stack=0x100000000:0x100800000 own=(nil) fp=0x40c45cd7), for accessing 0x40c45cd7 (code=1/prot=0), db=(nil)((nil):(nil)/(nil):(nil)/???:clean, hash:0/0) handler=(nil)
RSP-0x20:0x00000000004f087c RSP-0x18:0x00000000407ca0b0 RSP-0x10:0x00000000408756d0 RSP-0x08:0x00000001007fe5e0
RSP+0x00:0x00000000004f70e0 RSP+0x08:0x0000000000000002 RSP+0x10:0x0000000000000000 RSP+0x18:0x00000001007fe6e0
RAX:0x0000000000000000 RCX:0x0000000040c45cd6 RDX:0x00000000004f71d0 RBX:0x0000000000000003
RSP:0x00000001007fe548 RBP:0x0000000040c45cd7 RSI:0x0000000000000000 RDI:0x0000000040c45d5f
 R8:0x0000000000000000  R9:0x00000001007fe0cc R10:0x0000000000000000 R11:0x0000000000001002
R12:0x0000000040c45d5f R13:0x0000000040c45e48 R14:0x0000000040c45e78 R15:0x00000001007fe7b0
ES:0x002b CS:0x0033 SS:0x002b DS:0x002b FS:0x0043 GS:0x0053  x86opcode=C3 00 00 00 00 00 00 00
fish: Job 1, 'sudo -H ~/git/box64/build/box64…' terminated by signal SIGBUS (Misaligned address error)
```

**Comment by ptitSeb** [comment]
Mmmm, misalligned pthread stuff :(

Can you run with `BOX64_ROLLING_LOG=1` so we can see what function it's calling at the time of the issue?

**Comment by kaysond** [comment]
> Mmmm, misalligned pthread stuff :(
> 
> Can you run with `BOX64_ROLLING_LOG=1` so we can see what function it's calling at the time of the issue?

```
Last calls
194199|0x540130: Calling pthread_self (/lib/aarch64-linux-gnu/libc.so.6)(0xDB06A8, 0xBB, 0x653CB0F0, ...) => return 0x7FB6392E40
194199|0x540140: Calling pthread_mutex_lock (/lib/aarch64-linux-gnu/libpthread.so.0)(0xDB06A8, 0xBB, 0x653CB0F0, ...) => return 0x0
194199|0x540b59: Calling __clock_gettime (/lib/aarch64-linux-gnu/libc.so.6)(0x0, 0x1007FE2F0, 0x64E097B0, ...) => return 0x0
194199|0x540198: Calling __pthread_mutex_unlock (/lib/aarch64-linux-gnu/libpthread.so.0)(0xDB06A8, 0x0, 0x4B75E2, ...) => return 0x0
194199|0x5b2019: Calling __write (/lib/aarch64-linux-gnu/libc.so.6)(0x4, 0x653CB0F0, 0xBB, ...) => return 0xBB
194199|0x5404d5: Calling pthread_self (/lib/aarch64-linux-gnu/libc.so.6)(0xDB06A8, 0xBB, 0xBB, ...) => return 0x7FB6392E40
194199|0x5404e8: Calling pthread_mutex_lock (/lib/aarch64-linux-gnu/libpthread.so.0)(0xDB06A8, 0xBB, 0xBB, ...) => return 0x0
194199|0x540c3a: Calling __clock_gettime (/lib/aarch64-linux-gnu/libc.so.6)(0x0, 0x1007FE2F0, 0x64E097B0, ...) => return 0x0
194199|0x540548: Calling __pthread_mutex_unlock (/lib/aarch64-linux-gnu/libpthread.so.0)(0xDB06A8, 0x6050BDD30B580, 0x6050BDD3605B9, ...) => return 0x0
194199|0x59348e: Calling free (veeamtransport)(0x653CB0F0, 0x6050BDD30B580, 0x6050BDD3605B9, ...) => return 0x1007FE440
194199|0x593496: Calling __pthread_mutex_unlock (/lib/aarch64-linux-gnu/libpthread.so.0)(0x64DFA7D0, 0x6050BDD30B580, 0x6050BDD3605B9, ...) => return 0x0
194199|0x4ee3df: Calling free (veeamtransport)(0x653CAD50, 0x6050BDD30B580, 0x6050BDD3605B9, ...) => return 0x1007FE590
194199|0x4f7400: Calling my_pthread_mutex_init (veeamtransport)(0x653CAA67, 0x0, 0x653CAA00, ...) => return 0x0
194199|0x4f7426: Calling my_pthread_create (veeamtransport)(0x653CAA8F, 0x0, 0x4F71D0, ...) => return 0x0
194199|0x4f70e0: Calling pthread_mutex_lock (/lib/aarch64-linux-gnu/libpthread.so.0)(0x653CAA07, 0x1, 0x4F71D0, ...) => return
194199|0x5b1ec7: Calling __errno_location (/lib/aarch64-linux-gnu/libc.so.6)(0xDB2CC0, 0xBB, 0x653CB0F0, ...) => return 0x7FB6393720
194199|SIGBUS @0x7fb63d06fc (???(/lib/aarch64-linux-gnu/libpthread.so.0+0x7fb63d06fc)) (x64pc=0x802b3/???:"???", rsp=0x1007fe3e8, stack=0x100000000:0x100800000 own=(nil) fp=0x653caa07), for accessing 0x653caa07 (code=1/prot=0), db=(nil)((nil):(nil)/(nil):(nil)/???:clean, hash:0/0) handler=(nil)
RSP-0x20:0x00000000004f087c RSP-0x18:0x0000000064e63d50 RSP-0x10:0x0000000064fa4010 RSP-0x08:0x00000001007fe480
RSP+0x00:0x00000000004f70e0 RSP+0x08:0x0000000000000002 RSP+0x10:0x0000000000000000 RSP+0x18:0x00000001007fe580
RAX:0x0000000000000000 RCX:0x00000000653caa06 RDX:0x00000000004f71d0 RBX:0x0000000000000003
RSP:0x00000001007fe3e8 RBP:0x00000000653caa07 RSI:0x0000000000000001 RDI:0x00000000653caa07
 R8:0x00000001007fdf6c  R9:0x00000001007fdf6c R10:0x0000000000000000 R11:0x0000000000001002
R12:0x00000000653caa06 R13:0x0000000000000001 R14:0x00000000653caba8 R15:0x00000001007fe650
ES:0x002b CS:0x0033 SS:0x002b DS:0x002b FS:0x0043 GS:0x0053  x86opcode=C3 00 00 00 00 00 00 00
fish: Job 1, 'sudo -E ~/git/box64/build/box64…' terminated by signal SIGBUS (Misaligned address error)
```

Let me know if you need anything else from the logs. Really appreciate the help btw. Just shot over a donation to support the project. A ton of people would use this to run veeam backup repos on ARM if we can get it working!

**Comment by kaysond** [comment]
@ptitSeb any ideas what's going on here?

**Comment by ptitSeb** [comment]
Yeah, I haven't worked on that yet. I'm still perpexed by that log and don't see the logic behind that mutex. I'll probably have to try on my side if possible, to debug the issue.

**Comment by kaysond** [comment]
> Yeah, I haven't worked on that yet. I'm still perpexed by that log and don't see the logic behind that mutex. I'll probably have to try on my side if possible, to debug the issue.

Sure. I've uploaded everything here - https://mega.nz/file/Bx4W2IiD#SLuRWJOFsXtfToZQC8b58qgX408LlVlhBcuGszQ2Owk
You can also get them by installing veeam community edition and setting up a linux backup repo.

**Comment by kaysond** [comment]
Bumping :)

**Comment by ptitSeb** [comment]
Ah yeah, sorry, I did other stuffs. I'm currently working on threads, so it seem perfect time to revisit this one, I downloaded the binary and will start analysing the issue soon-ish. (also, just to note: that's the issue with continuing a ticket instead of closing / openning a new one: the title is no longer relevent to the current issue).

**Comment by kaysond** [comment]
> Ah yeah, sorry, I did other stuffs. I'm currently working on threads, so it seem perfect time to revisit this one, I downloaded the binary and will start analysing the issue soon-ish. (also, just to note: that's the issue with continuing a ticket instead of closing / openning a new one: the title is no longer relevent to the current issue).

Thanks for taking a look. Please let me know if you run into any issues running.

I can open a new issue or change the title if you'd like. LMK
